### ISSUE TITLE
Customize bibtex key based on user preferences

### ISSUE THREAD

**Issue by aaccomazzi** [adsabs/export_service#93]
We have heard from enough people that they want to be able to use a more mnemonic key for identifying a particular bibtex entry, rather than a bibcode.  Typically people would want the key to contain the first author's last name, then some combination of publication year, journal, additional authors, etc.  I think the most flexible thing we can do is allow a user to pick the key generation algorithm by specifying a custom format for it, with some pre-canned format easily selectable.

To provide a specific example, this is what we currently emit for `2019AAS...23338108A`:
```
@INPROCEEDINGS{2019AAS...23338108A,
       author = {{Accomazzi}, Alberto and {Kurtz}, Michael J. and {Henneken}, Edwin and
...
}
```
But instead I may want to have the key (currently the bibcode) be one of the following:
```
Accomazzi:2019              -- %1H:%Y
Accomazzi:2019:AAS          -- %1H:%Y:%q
Accomazzi2019               -- %1H%Y
Accomazzi2019AAS            -- %1H%Y%q
```
People could obviously opt to change the number of authors used to create a key as they see fit.  In all cases I would downgrade to ascii, remove whitespace and punctuation to make the key a single string, e.g.
```
AccomazziKurtz2019          -- %2H%Y
```
Which reminds me: we should probably have an option in the custom format to downgrade to ASCII (maybe %ZEncoding:ASCII?)

**Comment by hsgg** [comment]
This is a great first suggestion. Concerning the uniqueness problem, I will stipulate that it is important to have uniqueness within a custom-created library. Changing the number of authors will not always help.

The obvious solution is to append a lower-case letter to the year, e.g. `Accomazzi:2019a`. However, when adding more references to a library, the identifier may change to `Accomazzi:2019b`, and it would be necessary to update the citations in the paper, or else they will be pointing to the wrong reference.

Another solution may be to append the month or, if that is insufficient, the original ADS identifier *as needed* until the identifiers within the library are unique. If an identifier changes, then LaTeX will at least warn me. Would that be feasible?

That still has a failure mode when you substitute one reference with another and the mnemonic identifiers happen to be the same. I think that is acceptable, though.

**Comment by aaccomazzi** [comment]
I think this is not only reasonable, but also doable at least for libraries and/or search results up to a certain size.  What our code would do is generate preliminary keys for all records being exported, then analyze whether there is a duplication issue and if so proceed by appending a disambiguating letter to each identifier.

People who maintain libraries to be used as bibliography lists would simply download a fresh new copy of the library on a regular basis which would ensure uniqueness of the keys on an ongoing basis.  It's true that text that might have been written in the past referencing older keys would no longer be able to resolve the identifiers in the new system, but I would argue that that is the price to pay for having easy mnemonics as keys.

**Comment by golnazads** [comment]
aaccomazzi [10:57 AM]
So calling it "ASCII encoding" is actually a misnomer.  What I meant was that at least for the case in question we'd like an option to downgrade a unicode string to ASCII, at least in the case of generating a mnemonic bibtex key.  What I recommend using is the python `unidecode` package for this: https://pypi.org/project/Unidecode/
Now the question is: should we support this downgrade as a general formatting option for the export service?  We don't have many use cases to justify that, but if we are implementing it the way we describe in the ticket we have at least one.

**Comment by hsgg** [comment]
I was thinking of this again... I think my preferred option would be any of

    Accomazzi:2019AAS...23338108A
    Accomazzi+:2019AAS...23338108A
    Accomazzi/+:2019AAS...23338108A
    Accomazzi/Kurtz/+:2019AAS...23338108A

where the last bit after the colon is the original ADS identifier. This way there is no uniqueness problem. When writing a paper you can start typing the author's last name, and then any self-respecting modern text editor will be able to complete it for you (e.g. using `Ctrl-x Ctrl-o` in insert mode with vimtex).

**Comment by golnazads** [comment]
Try this format
```
@%W{%1h:%R,\n \t Author={%\A},\n \t Year = {%Y},\n \t Title = {%T},\n\t Abstract =\n{%\B},\n \t Journal = {%j},\n \t Volume = %V,\n\t Pages = {%pp},\n \t Url = {%u}\n}\n
```
which outputs in the format
```
@abstract{Henneken:2019LPI....50.1569H,
      Author={Henneken, E.~A., Muench, G., Holm Nielsen, L., Blanco-Cuaresma, S., and Accomazzi, A.},
      Year = {2019},
      Title = {Capturing Software Citations in Astronomy and Planetary Sciences},
     Abstract =
{This poster discusses the importance of capturing software citations and how this has been implemented in the Asclepias project of the AAS, ADS, and Zenodo.},
      Journal = {Lunar and Planetary Science Conference},
      Volume =
     Pages = {1569},
      Url = {https://ui.adsabs.harvard.edu/abs/2019LPI....50.1569H}
}
```
The first example. There is an issue that if there is a missing field value (i.e., volume here) it does not remove that text `Volume=`. I shall fix this issue in the next week or so.

**Comment by hsgg** [comment]
Thanks. Two issues with this: First, the first author's last name may be cut off if it contains a space. Second, the author list is not properly formatted. It is missing curly braces around the last names and the word `and` between names. Here is an example:

    @article{Di:2019JCAP...04..053D,
      Author={Di Dio, Enea, Durrer, Ruth, Maartens, Roy, Montanari, Francesco, and Umeh, Obinna},

**Comment by golnazads** [comment]
After I posted, Alberto brought my attention to the biggest problem going
forward which is that bibtex output is different for different types of
document. For example for document type @BOOK there is no keyword, or
pages, or month, while @ARTICLE has all these three fields. Hence one
custom format would not work and so various document types can not be
exported at one time. Alberto's thinking has been to include the id format
as a preference in user preferences section and have export apply that
custom format to the BibTex output the default being %R which is the
bibcode specifier.

Thank you so much for pointing the problem with multi part last name
without a hyphen. The this is a bug that I shall fix in the next day or two.

**Comment by golnazads** [comment]
supporting the following formats: %R, %nH (where n can be any number), %q, and %Y, respectively for bibcode, author, publication, and year.

released to dev on 6/4 and Build #261
